Adding Cloudstack Providor

syed opened the pull request:
This PR adds Cloudstack as a providor to PerfKitBenchmarker. Please review and let me know. 

Thanks,
-Syed

voellm commented:
Very cool thanks! Assigning to @ehankland  for help you integrate it.

voellm commented:
@syed   One thing we need is for you to sign the CLA.  You can find the instructions here - https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/CONTRIBUTING.md

This basics are you are contributing the code to OpenSource and the code is yours to contribute.

syed commented:
I signed it!

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_network.py:
```diff
@@ -0,0 +1,142 @@
+FLAGS = flags.FLAGS
+
+
+class CloudStackFirewall(network.BaseFirewall):
```
When you pull in the latest stuff from master, you shouldn't need this any more

syed replied:
Do you want me to rebase from the master?

voellm replied:
Yes please rebase from master.  All development is now done directly off
master.  We are going to delete the dev branch here soon.

On Wed, Oct 14, 2015 at 7:27 AM, Syed Mushtaq Ahmed <
notifications@github.com> wrote:

> In perfkitbenchmarker/cloudstack/cloudstack_network.py
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/527#discussion_r42000047>
> :
>
> > +flags.DEFINE_string('cs_network_offering',
> > +                    'DefaultIsolatedNetworkOfferingForVpcNetworksNoLB',
> > +                    'Name of the network offering')
> > +
> > +flags.DEFINE_string('cs_vpc_offering',
> > +                    'Default VPC offering',
> > +                    'Name of the VPC offering')
> > +
> > +flags.DEFINE_boolean('cs_use_vpc', True,
> > +                     'Use VPC to create networks')
> > +
> > +
> > +FLAGS = flags.FLAGS
> > +
> > +
> > +class CloudStackFirewall(network.BaseFirewall):
>
> Do you want me to rebase from the master?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/527/files#r42000047>
> .
>



-- 
Anthony F. Voellm (aka Tony)
Google Voice:  (650) 516-7382
<https://www.google.com/voice/b/0?pli=1#phones>
Blog: http://perfguy.blogspot.com/
Benchmark like a pro... PerfKitBenchmarker
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker>

ehankland commented on README.md:
````diff
@@ -159,6 +159,27 @@ echo topsecretpassword > ~/.config/openstack-password.txt
+(`--cs_use_vpc`), also specify the VPC offering (`--cs_vpc_offering`).
+
+```bash
+./pkb.py --cloud=CloudStack --benchmarks=ping --cs_network_offering=DefaultNetowrkOffering
````
sp. Netowrk

ehankland commented on README.md:
````diff
@@ -341,6 +362,12 @@ $ ./pkb.py --cloud=DigitalOcean --machine_type=16gb --benchmarks=iperf
+## Example run on CloudStack
+
+```bash
+./pkb.py --cloud=CloudStack --benchmarks=ping --cs_network_offering=DefaultNetowrkOffering
````
sp. Netowrk

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_disk.py:
```diff
@@ -0,0 +1,135 @@
+class CloudStackDisk(disk.BaseDisk):
+  """Object representing a Cloudstack Disk."""
+
+  _lock = threading.Lock()
```
Is the csapi library not threadsafe? If not, having a class level lock that you acquire for all operations will probably slow things down unnecessarily.

syed replied:
I checked with the author of csapi. The library is threadsafe. I'll remove the locks

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_network.py:
```diff
@@ -0,0 +1,142 @@
+    zone = self.cs.get_zone(zone_name)
+    if zone:
+        self.zone_id = zone['id']
+        self.zone = zone_name
```
Instead of super(CloudStackNetwork, self).__init__() consider super(CloudStackNetwork, self).__init__(zone_name) so self.zone gets set to zone_name for you.

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_virtual_machine.py:
```diff
@@ -0,0 +1,217 @@
+
+
+  @classmethod
+  def SetVmSpecDefaults(cls, vm_spec):
```
We are changing the way defaults get set due to the config update that's out for review right now (https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/510). Assuming that gets merged before this, you should delete this method and update the perfkitbenchmarker/configs/default_config_constants.yaml file. You can still set the default image in the __init__ method so that it will do the right thing automatically if people use different os_types.

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_virtual_machine.py:
```diff
@@ -0,0 +1,217 @@
+    """
+    super(CloudStackVirtualMachine, self).__init__(vm_spec)
+
+    self.network = \
```
will need to change when you pull in the latest updates (networks and firewalls now get passed in through __init__)

ehankland commented on perfkitbenchmarker/benchmark_spec.py:
```diff
@@ -116,6 +119,13 @@ def UnPickleLock(locked, *args):
+            DEBIAN: cs_vm.DebianBasedCloudStackVirtualMachine,
+            RHEL: cs_vm.CloudStackVirtualMachine
+        },
+        FIREWALL: cs_nw.CloudStackFirewall
```
When you merge with master you'll want to delete this, but add NETWORK: cs_nw.CloudStackNetwork

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_network.py:
```diff
@@ -0,0 +1,142 @@
+    if FLAGS.cs_use_vpc:
+
+        assert FLAGS.cs_vpc_offering, "VPC flag should specify the VPC offering"
+        vpc_off = self.cs.get_vpc_offering(FLAGS.cs_vpc_offering)
```
You are doing a good amount of work in this __init__ method (including multiple cs api calls). Consider moving the heavier work into a helper and calling it from Create

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_network.py:
```diff
@@ -0,0 +1,142 @@
+  def Delete(self):
+    """Deletes the actual network."""
+
+    if self.network_id:
```
I think this will break if the network doesn't get created successfully (because network_id only gets set at the end of Create) - maybe set self.network_id to None fairly early in the \_\_init\_\_ method

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_disk.py:
```diff
@@ -0,0 +1,135 @@
+    with self._lock:
+
+        res = self.cs.attach_volume(self.volume_id, vm.id)
+        assert res, "Unable to attach volume"
```
Any chance this error will be transient ? If so, may want to retry.

ehankland commented on perfkitbenchmarker/cloudstack/cloudstack_disk.py:
```diff
@@ -0,0 +1,135 @@
+    with self._lock:
+        self.cs.detach_volume(self.volume_id)
+
+  def GetDevicePath(self):
```
Shouldn't need this after the config PR goes in.

syed commented:
Thanks for your review @ehankland I'll make the required changes soon.

syed commented:
rebased with master and addressed @ehankland 's comments. Can you review this please? It looks very ugly though with all the commits. Sorry about that

ehankland commented:
I tried to clean up the commits. I pushed the result here: https://github.com/ehankland/PerfKitBenchmarker/tree/dev
You should be able to add that repository as a remote and update this PR by checking out the branch and pushing it to your remote. The commits still correctly attribute you as the author. Let me know if you need any help. Also please double check my work to make sure nothing got left out somehow.

In the future, you should be able to keep the commits cleaner by running git pull --rebase remote_name branch_name - This will undo your commits, pull down changes, and reapply your commits on top. If there are any conflicts you can fix them then continue the rebase.

syed commented:
I'll close this PR and open a new one which just has my changes rebased with the master so that it is easier to review for you.

ehankland commented:
Thanks
